### PR TITLE
REGRESSION: CMS_final: do not ignore CMS_dataFinal result

### PR DESCRIPTION
**Pull request by alonbl**
The CMS_dataFinal result is important as signature may fail, however, it
is ignored while returning success from CMS_final.

The SMIME_crlf_copy result is returned while it is ignored in the rest
of the openssl code.

Prefer the CMS_dataFinal result.

This is regression from openssl-1.

Blame finds c1669f41ei as root cause of regression.

Signed-off-by: Alon Bar-Lev <alon.barlev@gmail.com>

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by alonbl** [comment]
@mattcaswell please review regression from https://github.com/openssl/openssl/pull/11884

File: crypto/cms/cms_smime.c [hunk 1]
```diff
@@ -882,14 +882,17 @@ int CMS_final(CMS_ContentInfo *cms, BIO *data, BIO *dcont, unsigned int flags)
         return 0;
     }
 
-    ret = SMIME_crlf_copy(data, cmsbio, flags);
+    SMIME_crlf_copy(data, cmsbio, flags);
```
Comment (t8m) [inline crypto/cms/cms_smime.c:885]: The return value should be checked.

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/cms/cms_smime.c (hunk 1 above)
Comment (alonbl) [reply crypto/cms/cms_smime.c:885]: `git grep SMIME_crlf_copy` shows no check for return value all over, do you want to fix this in a separate patch for all locations or just this one? the only fail path for this is the BIO_new

**Comment by t8m** [comment]
We would need a signed CLA to be able to accept this if the check is added. See https://www.openssl.org/policies/cla.html

**Comment by alonbl** [comment]
> We would need a signed CLA to be able to accept this if the check is added. See https://www.openssl.org/policies/cla.html

Doesn't it falls in the *trivial* definition?

**Comment by t8m** [comment]
Without adding the check for the return value I suppose this could be acceptable with CLA: trivial. We can handle the check separately.

**Comment by alonbl** [comment]
@t8m: I added a new commit to handle the `SMIME_crlf_copy` status in all usages.

File: crypto/cms/cms_smime.c (hunk 1 above)
Comment (alonbl) [reply crypto/cms/cms_smime.c]: Added a separate commit to handle the `SMIME_crlf_copy` result code cross project

**Comment by paulidale** [comment]
The first commit would count as trivial IMO, both together wouldn't.

**Comment by alonbl** [comment]
> The first commit would count as trivial IMO, both together wouldn't.

In any case, I sent the cla :)

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by hlandau** [comment]
Merged to master and 3.0. Thank you.

**Comment by alonbl** [comment]
Hi,
Thank you for your prompt response.
Should I send pull request for openssl-3.0 branch or is it done automatically?
Thanks!

**Comment by adrien-n** [comment]
Hi,

We (Canonical/Ubuntu) would like to know the security implications of this issue or if this has not been analyzed. It looks like data that should fail authentication would be reported as succeeding but the lack of any mention of a security issue for this is surprising.

**Comment by t8m** [comment]
This function is used during the finalization step of creating CMS signed or encrypted messages, not when decrypting or verifying signatures.

It means some broken message that will later fail parsing/decryption/signature verification can be produced.

I do not see any security implications here.

**Comment by adrien-n** [comment]
I see, thanks a lot for your answer. I'll be discussing this with the
reporter of the issue on our side to better understand their need.

**Pull request closed without merging**